hashes/src/chacha20poly1305/mod.rs

```diff
@@ -0,0 +1,169 @@
+    /// The 16-byte authentication tag.
+    pub fn encrypt(self, content: &mut [u8], aad: Option<&[u8]>) -> [u8; 16] {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 1);
+        chacha.apply_keystream(content);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        poly.add(aad);
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - aad_overflow)]);
+        }
+
+        poly.add(content);
+        let text_overflow = content.len() % 16;
+        if text_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - text_overflow)]);
+        }
+
+        let aad_len = aad.len().to_le_bytes();
+        let msg_len = content.len().to_le_bytes();
+        let mut len_buffer = [0u8; 16];
+        len_buffer[..aad_len.len()].copy_from_slice(&aad_len[..]);
+        for i in 0..msg_len.len() {
+            len_buffer[i + aad_len.len()] = msg_len[i]
+        }
+        poly.add(&len_buffer);
+        poly.tag()
+    }
+
+    /// Decrypt the ciphertext in place if authentication tag is correct.
+    ///
+    /// # Arguments
+    ///
+    /// - `content` - Ciphertext to be decrypted in place.
+    /// - `tag`     - 16-byte authentication tag.
+    /// - `aad`     - Optional metadata covered by the authentication tag.
+    pub fn decrypt(
+        self,
+        content: &mut [u8],
+        tag: [u8; 16],
+        aad: Option<&[u8]>,
+    ) -> Result<(), Error> {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 0);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        poly.add(aad);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - aad_overflow)]);
+        }
+        poly.add(content);
+        let msg_overflow = content.len() % 16;
+        if msg_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - msg_overflow)]);
+        }
+
+        let aad_len = aad.len().to_le_bytes();
+        let msg_len = content.len().to_le_bytes();
+        let mut len_buffer = [0u8; 16];
+        len_buffer[..aad_len.len()].copy_from_slice(&aad_len[..]);
+        for i in 0..msg_len.len() {
+            len_buffer[i + aad_len.len()] = msg_len[i]
+        }
+        poly.add(&len_buffer);
+        let derived_tag = poly.tag();
+        if derived_tag.eq(&tag) {
+            let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 1);
+            chacha.apply_keystream(content);
+            Ok(())
+        } else {
+            Err(Error::UnauthenticatedAdditionalData)
+        }
+    }
+}
```

Comment:
Some kind of streamed decoding would be great too but I do acknowledge it's a potential footgun since requiring the check first can limit the exposure of vulnerabilities to authenticated peers.